### PR TITLE
Allow followed symlinks on walker

### PR DESCRIPTION
**Pull request by tonistiigi**
This allows setting paths on a walker that are evaluated as symlinks and the target paths on these paths are added to the includepatterns.

This enables fixing the cases where a Dockerfile context detection can make sure that if the source path that user requested is a symlink, the target path is included in the context as well.

### CODE REVIEW

File: walker.go [hunk 1]
```diff
@@ -199,29 +202,28 @@ func (s *StatInfo) Sys() interface{} {
...
-	pfxs := make([]string, 0, len(patterns))
-	for _, ptrn := range patterns {
-		idx := strings.IndexFunc(ptrn, func(ch rune) bool {
-			return ch == '*' || ch == '?' || ch == '[' || ch == '\\'
-		})
-		if idx == -1 {
-			idx = len(ptrn)
-		}
-		pfxs = append(pfxs, ptrn[:idx])
+func matchPrefix(pattern, name string) (bool, bool) {
```
Comment (thaJeztah) [inline walker.go]: Should this return a "match" for `matchPrefix("/hello/*", "/hello/foo/world")`?

File: walker.go [hunk 2]
```diff
@@ -39,8 +42,23 @@ func Walk(ctx context.Context, p string, opt *WalkOpt, fn filepath.WalkFunc) err
 		}
 	}
 
-	var lastIncludedDir string
 	var includePatterns []string
+	if opt != nil && opt.IncludePatterns != nil {
```
[the remark this answers is not in the document]
Comment (tonistiigi) [reply walker.go]: Without the followpaths update, there is no need to do this before walk is called. With the update it needs to be done before.

File: walker.go (hunk 1 above)
Comment (tonistiigi) [reply walker.go]: this is never reached as `hello/foo` would match and skip the whole directory

File: walker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply walker.go]: Ah, makes sense; I just noticed it was moved in the third commit 👍

**Comment by tonistiigi** [comment]
This is blocking https://github.com/moby/buildkit/pull/431

**Comment by tonistiigi** [comment]
@tiborvass updated

**Pull request merged**